src/coords.c

```diff
@@ -286,29 +287,36 @@ PyObject *GetCoords(PyArrayObject *arr)
         return NULL;
     }
     iternext = NpyIter_GetIterNext(iter, NULL);
-    if (iternext == NULL) { goto fail; }
+    if (iternext == NULL) { NpyIter_Deallocate(iter); return NULL; }
     dataptr = NpyIter_GetDataPtrArray(iter);
 
+    GEOS_INIT;
+
     /* We work with a "cursor" that tells the get_coordinates function where
     to write the coordinate data into the output array "result" */
     cursor = 0;
     do {
         /* get the geometry */
         obj = *(GeometryObject **) dataptr[0];
-        if (!get_geom(obj, &geom)) { goto fail; }
+        if (!get_geom(obj, &geom)) { errstate = PGERR_NOT_A_GEOMETRY; goto finish; }
         /* skip None values */
         if (geom == NULL) { continue; }
         /* get the coordinates */
-        if (!get_coordinates(context, geom, result, &cursor)) { goto fail; }
+        if (!get_coordinates(ctx, geom, result, &cursor)) { errstate = PGERR_GEOS_EXCEPTION; goto finish; }
     } while(iternext(iter));
 
-    NpyIter_Deallocate(iter);
-    return (PyObject *) result;
+    
+    finish:
 
-    fail:
-        Py_DECREF(result);
+        GEOS_FINISH;
         NpyIter_Deallocate(iter);
-        return NULL;
+
+        if (errstate == PGERR_SUCCESS ) {
```

Comment:
Suggestion: since both conditions return, make the bare return the successful one:
```
if (errstate != PGERR_SUCCESS) {
    Py_DECREF(result);
    return NULL;
}

return (PyObject *) result;
```